tests/lib/prepare-snapd.sh

```diff
@@ -0,0 +1,209 @@
+    PARAM_IMAGE="-drive file=${work_dir}/pc.img,cache=none,format=raw,id=disk1,if=none -device virtio-blk-pci,drive=disk1,bootindex=1"
+
+    SVC_NAME="nested-vm"
+    if ! sudo systemd-run --service-type=simple --unit="${SVC_NAME}" -- \
+                qemu-system-x86_64 \
+                ${PARAM_SMP} \
+                ${PARAM_CPU} \
+                ${PARAM_MEM} \
+                ${PARAM_TRACE} \
+                ${PARAM_LOG} \
+                ${PARAM_MACHINE} \
+                ${PARAM_DISPLAY} \
+                ${PARAM_NETWORK} \
+                ${PARAM_BIOS} \
+                ${PARAM_RANDOM} \
+                ${PARAM_IMAGE} \
+                ${PARAM_SERIAL} \
+                ${PARAM_MONITOR}; then
+        echo "Failed to start ${SVC_NAME}" 1>&2
+        systemctl status "${SVC_NAME}" || true
+        return 1
+    fi
+
+    # Wait until ssh is ready
+    if ! wait_for_ssh "${SVC_NAME}"; then
+        echo "===== SERIAL PORT OUTPUT ======" 1>&2
+        cat "${work_dir}/serial.log" 1>&2
+        echo "===== END OF SERIAL PORT OUTPUT ======" 1>&2
+        return 1
+    fi
+
+    # Wait for the snap command to become ready
+    nested_wait_for_snap_command
+}
+
+install_core22_deps() {
+    sudo apt update -qq
+
+    # these should already be installed in GCE and LXD images with the google/lxd-nested 
+    # backend, but in qemu local images from qemu-nested, we might not have them
+    sudo apt install psmisc fdisk snapd mtools ovmf qemu-system-x86 sshpass whois -yqq
+
+    # TODO: https://bugs.launchpad.net/snapd/+bug/1712808
+    # There is a bug in snapd that prevents udev rules from reloading in privileged containers
+    # with the following error message: 'cannot reload udev rules: exit status 1' when installing
+    # snaps. However it seems that retrying the installation fixes it
+    if ! sudo snap install snapcraft --classic; then
+        echo "FIXME: snapcraft install failed, retrying"
+        sudo snap install snapcraft --classic
+    fi
+    sudo snap install ubuntu-image --classic
+}
+
+download_core22_snaps() {
+    local snap_branch="$1"
+
+    # get the model
+    curl -o ubuntu-core-amd64-dangerous.model https://raw.githubusercontent.com/snapcore/models/master/ubuntu-core-22-amd64-dangerous.model
+
+    # download neccessary images
+    snap download pc-kernel --channel=22/${snap_branch} --basename=upstream-pc-kernel
+    snap download pc --channel=22/${snap_branch} --basename=upstream-pc-gadget
+    snap download snapd --channel=${snap_branch} --basename=upstream-snapd
+}
+
+create_cloud_init_cdimage_config() {
+    local CONFIG_PATH=$1
+    cat << 'EOF' > "$CONFIG_PATH"
+#cloud-config
+datasource_list: [NoCloud]
+users:
+  - name: ubuntu
+    sudo: "ALL=(ALL) NOPASSWD:ALL"
+    lock_passwd: false
+    plain_text_passwd: 'ubuntu'
+  - name: external
+    sudo: "ALL=(ALL) NOPASSWD:ALL"
+    lock_passwd: false
+    plain_text_passwd: 'ubuntu'
+  - name: test
```

Comment:
Why do we need all these users? Is that because the tests require them? Some comment would be helpful.

Reply:
Yea good point here - unfortunately all of them are required by various stages of testing. 'external' is used by the external backend, and 'test' is used during setup by snapd. Now that I think about it I'm not sure 'ubuntu' is required, I'll just try to verify if they are all required still, and then update the comments

Reply:
I've updated this and added comments, I have removed 'ubuntu'